src/irmin/tree.ml

```diff
@@ -1940,13 +1947,38 @@ module Make (P : Private.S) = struct
               k (`Node h))
 
     let to_tree p = tree_of_proof p Fun.id
-
-    let of_keys tree keys =
-      (* The following [clear] gets rid of existing backend values stored in
-         [info.value] and [info.findv_cache], but it doesn't clear the ones
-         stored in [v]. *)
-      clear tree;
-      let+ () = Lwt_list.iter_s (fun k -> find_tree tree k >|= ignore) keys in
-      of_tree tree
   end
+
+  let produce_proof repo kinded_hash f =
+    let tree = import_no_check repo kinded_hash in
+    let+ tree_after = f tree in
+    (* Here, we build a proof from [tree] (on not from [tree_after]!)
```

Comment:
I think this is not working.

Consider that initial tree:
```
/foo/age = 0
/foo/version = 1
/bar/age = 2
/bar/version = 3
```

On which we apply that `f` function:
```
let f t0 =
	let increment = function None -> assert false | Some i -> Some (i + 1) in
	let* t1 = Store.Tree.update t0 [ "foo"; "age" ] increment in
	let* t2 = Store.Tree.update t0 [ "bar"; "age" ] increment in
	t2
```

#### Before creation of t1
```
t0-a:
{"Blinded_node":"af0505a618875c533867c2bfde752fe243df4cf2"}
```


#### Before creation of t2
```
t0-b:
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t1-b:
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}
```

#### After creation of t2
```
t0-c (= t0-b)
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t1-c:
    {"Node":[ ["bar",{"Node":[ ["age",{    "Blinded_contents":["da4b9237bacccdf19c0760cab7aec4a8359010b0",{}]}],
                               ["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t2-c:
    {"Node":[ ["bar",{"Node":[ ["age",{    "Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}],
                               ["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}
```

#### Resulting proof
```
p (= t0-c) (= t0-b):
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

{"before":{"Node":"af0505a618875c533867c2bfde752fe243df4cf2"},"after":{"Node":"f99c9fe814cb981d510563fc4807d4ec0aea2230"},"proof": ... }

```

The code: https://github.com/samoht/irmin/compare/refactor-proofs-3...Ngoguey42:problematic-produce-proof

Reply:
The loading of `bar` for `t2` is not reflected in `t0`

Reply:
Hum I'm not sure why the cache of `bar` is not shared between `t1-c` and `t0-c`. I'll investigate.